<a href="https://colab.research.google.com/github/xqcxqc/CSC448/blob/main/448_hw4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [724]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# for data set split
from sklearn.model_selection import train_test_split

#import the iris dataset
from sklearn.datasets import load_iris
# to use LogisticRegression model
from sklearn.linear_model import LogisticRegression

# to use support vector machine
from sklearn import svm

#Use sklearn to train a Neural Network (MLP Classifier) on the training set
from sklearn.neural_network import MLPClassifier

#Use sklearn to train a k-Neighbors Classifier
from sklearn.neighbors import KNeighborsClassifier


# Part 1: Load the dataset

In [603]:
# Load the dataset (load remotely, not locally)
iris  = load_iris()

# Output the first 15 rows of the data

#convert the sklearn.utils.bunch to dataframe
df = pd.DataFrame(iris.data, columns=iris.feature_names)

# insert the type column to compelete the data frame
types = iris['target']
df['Types'] = types

df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    Types  
0       0  
1       0  
2       0  
3       0  
4       0  
5       0  
6       0  
7       0  
8       0  
9       0  
10      0  
11      0  
12      0  
13      0  
14      0

In [604]:
# Display a summary of the table information (number of datapoints, etc.)
print( df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   Types              150 non-null    int32  
dtypes: float64(4), int32(1)
memory usage: 5.4 KB
None


In [723]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)       Types  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?
<br>

This data set is about the iris flower. For each object in the dataser, it has five attributes which are "sepal length", "sepal width", "petal length", "petal width", and "types". These length and width values are quantitative(numeric) attributes and indicate the object's length and width. The type value is a categorical attribute and it tells what kind of flower the object is.
<br>

The features are "sepal length", "sepal width", "petal length", and petal width. The label is the "type". We map those features to the label. When the value of type is 0, the class of flower is <b>Setosa</b>. When the value of type is 1, the class of flower is <b>Versicolour</b>. When the value of type is 2, the class of flower is <b>Virginica</b>. 


# Part 2: Split the dataset into train and test

In [605]:
# Take the dataset and split it into our features (X) and label (y)

features = df[["sepal length (cm)", "sepal width (cm)","petal length (cm)", "petal width (cm)"]]
#print(features.head())

label = df["Types"]
#print(label.head(140))


# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train,X_test, y_train, y_test = train_test_split(features, label, test_size=0.1)
print("Features:")
print("# of training set: ", X_train.shape[0])
print("# of test set: ", X_test.shape[0])


print("Lable:")
print("# of training set: ", y_train.shape[0])
print("# of test set: ", y_test.shape[0])



Features:
# of training set:  135
# of test set:  15
Lable:
# of training set:  135
# of test set:  15


# Part 3: Logistic Regression

In [606]:
# i. Use sklearn to train a LogisticRegression model on the training set
clf = LogisticRegression(max_iter=1000).fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sampledata = np.array( [[6.0, 4.0, 1.2, 0.2]] )

probability = pd.DataFrame(clf.predict_proba(sampledata), columns=iris['target_names'])
probability


setosa  versicolor     virginica
0  0.986533    0.013467  5.045770e-09

In [621]:
#iii. Report on the score for Logistic regression model, what does the score measure? <br>
score = clf.score(X_test, y_test)
print("The score on testing set is", score)

score = clf.score(X_train, y_train)
print("The score on training set is", score)

The score on testing set is 0.9333333333333333
The score on training set is 0.9777777777777777


### Report:
The socre here is the mean accuracy on the testing data and label. The socre on the testing set is 0.93 out of 1. We can say that generlly the results from our model have about 7% of error to reach the exact result. Or we may say that the square of residual sum is 7% of the total sum of squares, which is small. We also can say that the model has 93% of chance to get the correct result. Therefore, we can think this score is good, and our model is able to work at most of time.

In [608]:
# iv. Extract the coefficents and intercepts for the boundary line(s)

print("The coefficents are ",clf.coef_)
print("The intercepts are ",clf.intercept_)

The coefficents are  [[-0.44335822  0.8820754  -2.43760004 -1.03732152]
 [ 0.3219673  -0.32763052 -0.19819612 -0.8871256 ]
 [ 0.12139093 -0.55444488  2.63579616  1.92444712]]
The intercepts are  [ 10.05946055   3.29507865 -13.3545392 ]


# Part 4: Support Vector Machine

In [797]:
# i. Use sklearn to train a Support Vector Classifier on the training set

    # Support vector machines (SVMs) are a set of supervised learning methods used for 
    # classification, regression and outliers detection.

svm_clf = svm.SVC(probability = True, random_state = np.random.seed())
    # probabilitybool, default=False
    # Whether to enable probability estimates. This must be enabled prior to calling fit, 
    # will slow down that method as it internally uses 5-fold cross-validation, 
    # and predict_proba may be inconsistent with predict.
svm_clf.fit(X_train,y_train)

SVC(probability=True)

In [798]:
# ii. For a sample datapoint, predict the probabilities for each possible class

sampledata = np.array( [[6.0, 4.0, 1.2, 0.2]] )

probability = pd.DataFrame(svm_clf.predict_proba(sampledata), columns=iris['target_names'])
probability


setosa  versicolor  virginica
0  0.973939    0.015288   0.010773

In [799]:
# iii. Report on the score for the SVM, what does the score measure?
svm_score = svm_clf.score(X_test, y_test)
print("The score on testing set is", svm_score)


The score on testing set is 0.9333333333333333


### Report:
 The socre on the testing set is 0.93 out of 1. We can say that generlly the results from our model have about 7% of error to reach the exact result, which is small. We also can say that the model has 93% of chance to get the correct result. Therefore, we can think this score is good, and our model is able to work at most of time.

# Part 5: Neural Network

In [618]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
nn_clf = MLPClassifier(solver='lbfgs',max_iter=1000,
                       hidden_layer_sizes=(5,4), activation="logistic", random_state=np.random.seed())

nn_clf.fit(X_train, y_train)


MLPClassifier(activation='logistic', hidden_layer_sizes=(5, 4), max_iter=1000,
              solver='lbfgs')

In [619]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sampledata = np.array( [[6.0, 4.0, 1.2, 0.2]] )

probability = pd.DataFrame(nn_clf.predict_proba(sampledata), columns=iris['target_names'])
probability

setosa    versicolor     virginica
0     1.0  5.312810e-08  6.696909e-15

In [620]:
# iii. Report on the score for the Neural Network, what does the score measure?
nn_score = nn_clf.score(X_test, y_test)
print("The score on testing set is", nn_score)


The score on testing set is 1.0


### Report:
The socre here is the mean accuracy on the testing data and label. The socre on the testing set is 1 out of 1. We can say that the results from our model have no error, or the model has 100% of chance to get the correct result. Therefore, we can think this score is perfect.

In [719]:
# iv: Experiment with different options for the neural network, 
#report on your best configuration 
#(the highest score I was able to achieve was 0.8666)

sgd = MLPClassifier(solver='sgd',max_iter=3000,
                       hidden_layer_sizes=(7), activation="logistic", random_state=np.random.seed())

sgd.fit(X_train, y_train)

sgd_score = sgd.score(X_test, y_test)

print("Using stochastic gradient descent solver, the score on testing set is", sgd_score)


Using stochastic gradient descent solver, the score on testing set is 0.8666666666666667


In [722]:
adam = MLPClassifier(solver='adam',max_iter=3000,
                       hidden_layer_sizes=(4), activation="logistic", random_state=np.random.seed())

adam.fit(X_train, y_train)

adam_score = adam.score(X_test, y_test)

print("Using adam solver, the score on testing set is", adam_score)

Using adam solver, the score on testing set is 1.0


## Report:
<br>
To experiment with different options for the neural network, I use all 3 types of slover.<br><br>
For "lbfgs" slover, the configuration I use is (5,4) for the hidden layer size. And the activation fucntion is logistic function(sigmoid funciton). The score on the testing set is 1, which is the perfect socre and the best score I got. However, there is small chance to get other scores like 0.866 and 0933, when we run the code many times. 
<br><br>
The second solver is "sgd". the configuration I use is (7) for the hidden layer size. And the activation fucntion is logistic function. The score on the testing set is 0.866, which is the best score I got for this configuration. However, there is small chance to get other scores like 0.533 and 0933, when we run the code many times. 
<br><br>
The last solver is "adam". In this time, the configuration I use is (4) for the hidden layer size. And the activation fucntion is logistic function. The score on the testing set is 1, which is the best score I got for this configuration. This configuration is more stable than the other two, because the socre is the same while I run the code many times. 
<br>
I think the best configuration I have is the one using "adam" solver. Because the socre is the perfect score, and I think this configuraion is the most stable one.

# Part 6: K-Nearest Neighbors

In [794]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

knn_clf = KNeighborsClassifier(n_neighbors=28)
knn_clf.fit(X_train, y_train)


KNeighborsClassifier(n_neighbors=28)

In [795]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sampledata = np.array( [[6.0, 4.0, 1.2, 0.2]] )

probability = pd.DataFrame(knn_clf.predict_proba(sampledata), columns=iris['target_names'])
probability


setosa  versicolor  virginica
0     1.0         0.0        0.0

In [796]:
# iii. Report on the score for kNN, what does the score measure?
knn_score = knn_clf.score(X_test, y_test)
print("The score on testing set is", knn_score)


The score on testing set is 0.9333333333333333


## Report:
 The socre on the testing set is 0.93 out of 1. We can say that generlly the results from our model have about 7% of error to reach the exact result, which is small. We also can say that the model has 93% of chance to get the correct result. Therefore, we can think this score is good, and our model is able to work at most of time.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

## Report:
<br>
In general, the results of these models are all good. Most of them have scores higher than 90%, which means that these models are able to predict correctly in the most of times. 
<br> <br> 
One thing surprises me is that for K-Nearest Neighbors model, the score decreases when the neighbors is larger. According to my test, this model's socre is 0.933 when the number of neighbors is from 1 to 28. After that, the socre decrease when the number of neighbors reach a bigger enough number. This suprises me because I learn that it might not be good if k is large. And it could be hard to use, since we need to choose the best k for different dataset.
<br> <br> 
The other thing surprise me is the Neural Network model. The socres I got from the same configuration could be different sometimes when I used lbfgs and sgd as my solver. And one more thing surprise me is adam solver. Compare to lbfgs and sgd, adam solver gives me the perfect score even it works well on large dataset(our dataset is small). And according to my experiments, the score I got from using adam solver is stable. 
<br><br>
Some of these model are really good such as Logistic Regression. Because it can give us high socre, and it is easy to implement.  However, I think Neural network with adam solver is the model performed the best on the dataset .
Because Neural network can give us the perfect score, which is the highest score here. One bad thing is that it take more iterations to converge, which make it slower than other models.